File: packages/custos-js/src/bundle.ts
import { createHash } from "node:crypto";
import { createReadStream, existsSync, mkdtempSync, readdirSync, readFileSync, rmSync, writeFileSync, statSync } from "node:fs";
import { createGzip, createGunzip } from "node:zlib";
import { tmpdir } from "node:os";
import { dirname, join, relative } from "node:path";
import { pipeline } from "node:stream/promises";
import { Readable, Writable } from "node:stream";

import { canonicalBytes } from "./canonical.js";
import { publicKeyFromB64, verifySignature, KeyPair } from "./keys.js";
import { verifyLedger, VerifyResult } from "./verify.js";

// Minimal ustar tar writer/reader — sufficient for our bundle format.

export interface TarEntry { name: string; data: Buffer; }

function tarPad(n: number): number { return (512 - (n % 512)) % 512; }

// Exported for internal tests only; NOT part of the public API.
export function _buildTarForTest(entries: TarEntry[]): Buffer {
  return buildTar(entries);
}
export function _readTarForTest(buf: Buffer): TarEntry[] {
  return readTar(buf);
}
export const _gzipBufferForTest = gzipBuffer;
export const _gunzipBufferForTest = gunzipBuffer;

function buildTar(entries: TarEntry[]): Buffer {
  const chunks: Buffer[] = [];
  for (const e of entries) {
    const header = Buffer.alloc(512, 0);
    header.write(e.name.slice(0, 100), 0, "utf8");
    header.write("0000644", 100, "ascii"); // mode
    header.write("0000000", 108, "ascii"); // uid
    header.write("0000000", 116, "ascii"); // gid
    header.write(e.data.length.toString(8).padStart(11, "0"), 124, "ascii"); // size
    header.write(Math.floor(Date.now() / 1000).toString(8).padStart(11, "0"), 136, "ascii"); // mtime
    header.write("        ", 148, "ascii"); // chksum placeholder
    header.write("0", 156, "ascii"); // typeflag
    header.write("ustar  ", 257, "ascii");
    // checksum
    let sum = 0;
    for (const b of header) sum += b;
    header.write(sum.toString(8).padStart(6, "0") + "\0 ", 148, "ascii");
    chunks.push(header, e.data, Buffer.alloc(tarPad(e.data.length), 0));
  }
  chunks.push(Buffer.alloc(1024, 0)); // end marker
  return Buffer.concat(chunks);
}

async function gzipBuffer(buf: Buffer): Promise<Buffer> {
  const chunks: Buffer[] = [];
  await pipeline(Readable.from(buf), createGzip(), new Writable({
    write(chunk, _enc, cb) { chunks.push(Buffer.from(chunk)); cb(); },
  }));
  return Buffer.concat(chunks);
}

async function gunzipBuffer(buf: Buffer): Promise<Buffer> {
  const chunks: Buffer[] = [];
  await pipeline(Readable.from(buf), createGunzip(), new Writable({
    write(chunk, _enc, cb) { chunks.push(Buffer.from(chunk)); cb(); },
  }));
  return Buffer.concat(chunks);
}

function readTar(buf: Buffer): TarEntry[] {
  const entries: TarEntry[] = [];
  let off = 0;
  while (off + 512 <= buf.length) {
    const header = buf.subarray(off, off + 512);
    if (header.every((b) => b === 0)) break;
    const nameEnd = header.indexOf(0, 0);
    const name = header.subarray(0, nameEnd >= 0 && nameEnd < 100 ? nameEnd : 100).toString("utf8");
    const sizeStr = header.subarray(124, 135).toString("ascii").replace(/\0/g, "").trim();
    const size = parseInt(sizeStr, 8);
    off += 512;
    const data = buf.subarray(off, off + size);
    entries.push({ name, data: Buffer.from(data) });
    off += size + tarPad(size);
  }
  return entries;
}

/**
 * Compute the ``policies_hash`` manifest value for a set of policy files.
 *
 * Algorithm (spec/WIRE.md §5, added in v0.3.0):
 *   1. Sort by basename lexicographically.
 *   2. For each file, sha256(bytes) as hex.
 *   3. Canonical JSON of [{name, sha256}, ...].
 *   4. sha256 of that canonical bytes, hex-encoded, prefixed "sha256:".
 */
function computePoliciesHash(files: { name: string; data: Buffer }[]): string {
  const entries = files
    .map((f) => ({ name: f.name, sha256: createHash("sha256").update(f.data).digest("hex") }))
    .sort((a, b) => (a.name < b.name ? -1 : a.name > b.name ? 1 : 0));
  const canonical = canonicalBytes(entries);
  return "sha256:" + createHash("sha256").update(canonical).digest("hex");
}

export async function createBundle(
  ledgerPath: string, pubkeyPath: string, output: string, kp: KeyPair, policiesDir?: string,
): Promise<string> {
  const ledger = readFileSync(ledgerPath);
  const pub = readFileSync(pubkeyPath);
  let records = 0;
  for (const line of ledger.toString("utf8").split("\n")) if (line.trim()) records++;

  // Collect policy files up front so both the tar entries and the manifest
  // hash see the same bytes.
  //
  // v0.4.0: if no explicit policiesDir is passed, look for
  // `<ledger.parent>/policies/` — the default directory the Gate SDK
  // snapshots content-addressed policy files into. This preserves every
  // policy version referenced by any record in the ledger, not just the
  // latest. See spec/WIRE.md §5.1.
  //
  // Also: walk recursively with posix-relative names so the bundle layout
  // matches the Python writer byte-for-byte (spec/WIRE.md §5).
  let effectivePoliciesDir = policiesDir;
  if (!effectivePoliciesDir) {
    const defaultDir = join(dirname(ledgerPath), "policies");
    if (existsSync(defaultDir) && statSync(defaultDir).isDirectory()) {
      effectivePoliciesDir = defaultDir;
    }
  }
  const policyFiles: { name: string; data: Buffer }[] = [];
  if (effectivePoliciesDir && existsSync(effectivePoliciesDir) && statSync(effectivePoliciesDir).isDirectory()) {
    walkFilesSorted(effectivePoliciesDir, (full) => {
      const rel = relative(effectivePoliciesDir!, full).split(/[\\/]/).join("/");
      policyFiles.push({ name: rel, data: readFileSync(full) });
    });
  }

  const manifest: Record<string, unknown> = {
    v: 1,
    created: new Date().toISOString().replace(/\.\d+Z$/, "Z"),
    records,
    pubkey: pub.toString("ascii").trim(),
  };
  // Optional field added in v0.3.0; older verifiers ignore it, newer verifiers
  // enforce it. Not emitted when there are no policies (backwards compat).
  if (policyFiles.length > 0) {
    manifest.policies_hash = computePoliciesHash(policyFiles);
  }
  const manifestBytes = canonicalBytes(manifest);
  const digest = createHash("sha256").update(manifestBytes).digest();
  const sig = "ed25519:" + kp.sign(digest).toString("base64");

  const entries: TarEntry[] = [
    { name: "bundle/manifest.json", data: Buffer.from(manifestBytes) },
    { name: "bundle/manifest.sig", data: Buffer.from(sig, "ascii") },
    { name: "bundle/ledger.jsonl", data: ledger },
    { name: "bundle/ledger.pub", data: pub },
  ];
  for (const pf of policyFiles) {
    entries.push({ name: `bundle/policies/${pf.name}`, data: pf.data });
  }
  const tar = buildTar(entries);
  const gz = await gzipBuffer(tar);
  writeFileSync(output, gz);
  return output;
}

/**
 * Depth-first walk of `root`, invoking `onFile` for every regular file with
 * an absolute path. Entries at each directory level are visited in
 * lexicographic order for reproducibility.
 */
function walkFilesSorted(root: string, onFile: (fullPath: string) => void): void {
  const stack: string[] = [root];
  while (stack.length > 0) {
    const cur = stack.pop()!;
    const names = readdirSync(cur).sort();
    for (const name of names) {
      const full = join(cur, name);
      const st = statSync(full);
      if (st.isDirectory()) stack.push(full);
      else if (st.isFile()) onFile(full);
    }
  }
}

export async function verifyBundle(path: string): Promise<VerifyResult & { manifest?: unknown }> {
  const gz = readFileSync(path);
  const tar = await gunzipBuffer(gz);
  const entries = readTar(tar);
  const byName = new Map(entries.map((e) => [e.name, e.data]));
  const manifestBytes = byName.get("bundle/manifest.json");
  const sig = byName.get("bundle/manifest.sig")?.toString("ascii").trim();
  const ledger = byName.get("bundle/ledger.jsonl");
  const pubB64 = byName.get("bundle/ledger.pub")?.toString("ascii").trim();
  if (!manifestBytes || !sig || !ledger || !pubB64) return { ok: false, records: 0, errors: ["bundle incomplete"] };
  const manifest = JSON.parse(manifestBytes.toString("utf8"));
  const pub = publicKeyFromB64(pubB64);
  const digest = createHash("sha256").update(canonicalBytes(manifest)).digest();
  if (!sig.startsWith("ed25519:")) return { ok: false, records: 0, errors: ["manifest sig format invalid"] };
  if (!verifySignature(pub, Buffer.from(sig.slice(8), "base64"), digest)) {
    return { ok: false, records: 0, errors: ["manifest signature invalid"] };
  }

  // Optional policies_hash enforcement (v0.3.0+). Missing → skip (old bundles).
  if (typeof (manifest as any).policies_hash === "string") {
    const policyFiles: { name: string; data: Buffer }[] = [];
    const prefix = "bundle/policies/";
    for (const e of entries) {
      if (e.name.startsWith(prefix) && e.data.length > 0) {
        policyFiles.push({ name: e.name.slice(prefix.length), data: e.data });
      }
    }
    const recomputed = computePoliciesHash(policyFiles);
    if (recomputed !== (manifest as any).policies_hash) {
      return {
        ok: false,
        records: 0,
        errors: [`policies_hash mismatch: manifest=${(manifest as any).policies_hash} recomputed=${recomputed}`],
        manifest,
      };
    }
  }

  const td = mkdtempSync(join(tmpdir(), "custos-bundle-"));
  try {
    writeFileSync(join(td, "ledger.jsonl"), ledger);
    writeFileSync(join(td, "ledger.pub"), pubB64);
    const r = verifyLedger(join(td, "ledger.jsonl"), join(td, "ledger.pub"));
    return { ...r, manifest };
  } finally {
    rmSync(td, { recursive: true, force: true });
  }
}

File: packages/custos-py/src/custos/bundle.py
"""Portable evidence bundle: tar.gz containing ledger + pubkey + policy snapshot + signed manifest."""
from __future__ import annotations

import base64
import hashlib
import io
import json
import tarfile
import time
from pathlib import Path
from typing import Optional

from cryptography.exceptions import InvalidSignature

from custos.canonical import dumps as canonical_dumps
from custos.keys import KeyPair, public_key_from_b64


def _compute_policies_hash(policy_files: list[tuple[str, bytes]]) -> str:
    """Compute the ``policies_hash`` value for a set of (basename, bytes) pairs.

    Algorithm (see spec/WIRE.md §5):
      1. Sort by basename lexicographically.
      2. For each file, sha256(bytes).hexdigest().
      3. Canonical JSON of [{"name": <basename>, "sha256": <hex>}, ...].
      4. sha256 of that canonical bytes, hex-encoded, prefixed with "sha256:".
    """
    entries = sorted(
        (
            {"name": name, "sha256": hashlib.sha256(data).hexdigest()}
            for name, data in policy_files
        ),
        key=lambda e: e["name"],
    )
    canonical = canonical_dumps(entries)
    return "sha256:" + hashlib.sha256(canonical).hexdigest()


def create_bundle(
    ledger_path: str | Path,
    pubkey_path: str | Path,
    output_path: str | Path,
    keypair: KeyPair,
    policies_dir: Optional[str | Path] = None,
) -> Path:
    ledger_path = Path(ledger_path)
    pubkey_path = Path(pubkey_path)
    output_path = Path(output_path)

    records = 0
    with ledger_path.open("rb") as f:
        for line in f:
            if line.strip():
                records += 1

    # Gather policy files (if any) so we can both embed them AND commit to
    # their contents via the manifest's optional ``policies_hash`` field.
    #
    # v0.4.0: if no explicit policies_dir is passed, look for
    # ``<ledger.parent>/policies/`` — the default directory the Gate SDK
    # snapshots content-addressed policy files into. This preserves every
    # policy version referenced by any record in the ledger, not just the
    # latest. See spec/WIRE.md §5.1.
    policy_entries: list[tuple[str, str, bytes]] = []  # (tar_name, basename, data)
    if policies_dir is None:
        default = ledger_path.parent / "policies"
        if default.is_dir():
            policies_dir = default
    if policies_dir:
        pd = Path(policies_dir)
        if pd.is_dir():
            for p in sorted(pd.rglob("*")):
                if p.is_file():
                    rel = p.relative_to(pd)
                    policy_entries.append(
                        (f"bundle/policies/{rel.as_posix()}", rel.as_posix(), p.read_bytes())
                    )

    manifest = {
        "v": 1,
        "created": time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime()),
        "records": records,
        "pubkey": Path(pubkey_path).read_bytes().decode("ascii").strip(),
    }
    # Added in v0.3.0 — optional; verifiers that see it MUST enforce it, but
    # bundles produced before this field existed remain valid.
    if policy_entries:
        manifest["policies_hash"] = _compute_policies_hash(
            [(name, data) for (_tn, name, data) in policy_entries]
        )
    manifest_bytes = canonical_dumps(manifest)
    digest = hashlib.sha256(manifest_bytes).digest()
    sig = keypair.sign(digest)
    sig_line = "ed25519:" + base64.b64encode(sig).decode("ascii")

    with tarfile.open(output_path, "w:gz") as tar:
        _add_file(tar, "bundle/manifest.json", manifest_bytes)
        _add_file(tar, "bundle/manifest.sig", sig_line.encode("ascii"))
        _add_file(tar, "bundle/ledger.jsonl", ledger_path.read_bytes())
        _add_file(tar, "bundle/ledger.pub", pubkey_path.read_bytes())
        for tar_name, _basename, data in policy_entries:
            _add_file(tar, tar_name, data)
    return output_path


def _add_file(tar: tarfile.TarFile, name: str, data: bytes) -> None:
    info = tarfile.TarInfo(name=name)
    info.size = len(data)
    info.mtime = int(time.time())
    tar.addfile(info, io.BytesIO(data))


def verify_bundle(path: str | Path) -> dict:
    """Return {ok, records, errors} after verifying manifest signature and ledger chain."""
    from custos.verify import verify_ledger

    path = Path(path)
    policy_members: list[tuple[str, bytes]] = []  # (basename, data)
    with tarfile.open(path, "r:gz") as tar:
        manifest_b = _read_member(tar, "bundle/manifest.json")
        sig_b = _read_member(tar, "bundle/manifest.sig").decode("ascii").strip()
        ledger_b = _read_member(tar, "bundle/ledger.jsonl")
        pub_b = _read_member(tar, "bundle/ledger.pub").decode("ascii").strip()
        # Collect any embedded policies for optional policies_hash check.
        for m in tar.getmembers():
            if m.isfile() and m.name.startswith("bundle/policies/"):
                f = tar.extractfile(m)
                if f is not None:
                    policy_members.append((m.name[len("bundle/policies/"):], f.read()))

    manifest = json.loads(manifest_b)
    pub = public_key_from_b64(pub_b)
    digest = hashlib.sha256(canonical_dumps(manifest)).digest()
    if not sig_b.startswith("ed25519:"):
        return {"ok": False, "records": 0, "errors": ["manifest sig format invalid"]}
    try:
        pub.verify(base64.b64decode(sig_b.split(":", 1)[1]), digest)
    except InvalidSignature:
        return {"ok": False, "records": 0, "errors": ["manifest signature invalid"]}

    # Optional policies_hash enforcement (added in v0.3.0). Absent → skip
    # (backwards compat with older bundles).
    if "policies_hash" in manifest:
        recomputed = _compute_policies_hash(policy_members)
        if recomputed != manifest["policies_hash"]:
            return {
                "ok": False,
                "records": 0,
                "errors": [
                    f"policies_hash mismatch: manifest={manifest['policies_hash']} "
                    f"recomputed={recomputed}"
                ],
                "manifest": manifest,
            }

    # Extract to temp files and reuse verifier
    import tempfile
    with tempfile.TemporaryDirectory() as td:
        tdp = Path(td)
        (tdp / "ledger.jsonl").write_bytes(ledger_b)
        (tdp / "ledger.pub").write_bytes(pub_b.encode("ascii"))
        r = verify_ledger(tdp / "ledger.jsonl", tdp / "ledger.pub")
    return {"ok": r.ok, "records": r.records, "errors": r.errors, "manifest": manifest}


def _read_member(tar: tarfile.TarFile, name: str) -> bytes:
    m = tar.getmember(name)
    f = tar.extractfile(m)
    assert f is not None
    return f.read()
